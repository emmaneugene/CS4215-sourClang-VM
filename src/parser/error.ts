import * as es from 'estree'

import { ErrorSeverity, ErrorType, SourceError } from '../types'
import { stripIndent } from '../utils/formatters'

export class DisallowedConstructError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public nodeType: string

  constructor(public node: es.Node) {
    this.nodeType = this.formatNodeType(this.node.type)
  }

  get location() {
    return this.node.loc!
  }

  public explain() {
    return `${this.nodeType} are not allowed`
  }

  public elaborate() {
    return stripIndent`
      You are trying to use ${this.nodeType}, which is not allowed (yet).
    `
  }

  /**
   * Converts estree node.type into english
   * e.g. ThisExpression -> 'this' expressions
   *      Property -> Properties
   *      EmptyStatement -> Empty Statements
   */
  private formatNodeType(nodeType: string) {
    switch (nodeType) {
      case 'ThisExpression':
        return "'this' expressions"
      case 'Property':
        return 'Properties'
      default: {
        const words = nodeType.split(/(?=[A-Z])/)
        return words.map((word, i) => (i === 0 ? word : word.toLowerCase())).join(' ') + 's'
      }
    }
  }
}

export class FatalSyntaxError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public constructor(public location: es.SourceLocation, public message: string) {}

  public explain() {
    return this.message
  }

  public elaborate() {
    return 'There is a syntax error in your program'
  }
}

export class MissingSemicolonError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public constructor(public location: es.SourceLocation) {}

  public explain() {
    return 'Missing semicolon at the end of statement'
  }

  public elaborate() {
    return 'Every statement must be terminated by a semicolon.'
  }
}

export class TrailingCommaError implements SourceError {
  public type: ErrorType.SYNTAX
  public severity: ErrorSeverity.WARNING
  public constructor(public location: es.SourceLocation) {}

  public explain() {
    return 'Trailing comma'
  }

  public elaborate() {
    return 'Please remove the trailing comma'
  }
}

export class InvalidConfigError implements SourceError {
  type: ErrorType
  severity: ErrorSeverity
  location: es.SourceLocation

  explain(): string {
    return 'Something went wrong while parsing'
  }

  elaborate(): string {
    return 'Please check the stacktrace'
  }
}

export class InvalidDataType implements SourceError {
  type: ErrorType
  severity: ErrorSeverity
  location: es.SourceLocation

  explain(): string {
    return 'This datatype is not supported'
  }
  elaborate(): string {
    return 'Please check the declared datatype'
  }
}
