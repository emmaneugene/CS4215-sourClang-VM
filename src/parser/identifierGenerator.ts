import { ErrorNode } from 'antlr4ts/tree/ErrorNode'
import { ParseTree } from 'antlr4ts/tree/ParseTree'
import { RuleNode } from 'antlr4ts/tree/RuleNode'
import { TerminalNode } from 'antlr4ts/tree/TerminalNode'
import * as es from 'estree'

import { DeclaratorContext } from '../lang/SourCParser'
import { SourCParserVisitor } from '../lang/SourCParserVisitor'
import { getPointerList } from './utils'

class IdentifierGenerator implements SourCParserVisitor<Omit<es.Identifier, 'datatype'>> {
  visitDeclarator(ctx: DeclaratorContext): Omit<es.Identifier, 'datatype'> {
    if (ctx._array) {
      // array
      throw new Error('Method not implemented')
    }

    if (ctx._funcPointer) {
      // function pointers
      throw new Error('Method not implemented')
    }

    if (ctx._funcProto) {
      // function prototypes
      // TODO: No support for function prototypes yet
      throw new Error('Method not implemented')
    }

    return {
      type: 'Identifier',
      name: ctx.Identifier().text,
      pointerList: getPointerList(ctx.pointer())
    }
  }

  visit(tree: ParseTree): Omit<es.Identifier, 'datatype'> {
    return tree.accept(this)
  }

  visitChildren(node: RuleNode): Omit<es.Identifier, 'datatype'> {
    throw new Error('Method not implemented.')
  }

  visitTerminal(node: TerminalNode): Omit<es.Identifier, 'datatype'> {
    return node.accept(this)
  }

  visitErrorNode(node: ErrorNode): Omit<es.Identifier, 'datatype'> {
    throw new Error('Method not implemented.')
  }
}

export default IdentifierGenerator
