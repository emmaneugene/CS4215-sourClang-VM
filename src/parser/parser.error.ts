import { SourceLocation } from 'estree'

export class FatalSyntaxError extends Error {
  constructor(loc: SourceLocation) {
    const { start, end } = loc
    const message =
      `Error around: ` +
      `Column: ${start.column}-${end.column}.` +
      `Line: ${start.line}-${end.line}`

    super(message)
    this.name = 'FatalSyntaxError'
    this.message = message
  }
}

export class InvalidConfigError extends Error {
  constructor(loc: SourceLocation) {
    const { start, end } = loc
    const message =
      `Error around: ` +
      `Column: ${start.column}-${end.column}.` +
      `Line: ${start.line}-${end.line}`

    super(message)
    this.name = 'InvalidConfigError'
    this.message = message
  }
}
