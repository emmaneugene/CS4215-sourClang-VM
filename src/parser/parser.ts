import { CharStreams, CommonTokenStream } from 'antlr4ts'
import * as es from 'estree'

import { SourCLexer } from '../lang/SourCLexer'
import { Context, ErrorSeverity } from '../types'
import { SourCParser, TranslationUnitContext } from './../lang/SourCParser'
import { FatalSyntaxError } from './error'
import StatementGenerator from './statementGenerator'

function convertStatement(translationUnit: TranslationUnitContext): es.Statement {
  const generator = new StatementGenerator()
  return translationUnit.accept(generator)
}

function convertSource(translationUnit: TranslationUnitContext): es.Program {
  return {
    type: 'Program',
    sourceType: 'script',
    body: [convertStatement(translationUnit)]
  }
}

export function parse(source: string, context: Context) {
  let program: es.Program | undefined

  if (context.variant === 'calc') {
    const inputStream = CharStreams.fromString(source)
    const lexer = new SourCLexer(inputStream)
    const tokenStream = new CommonTokenStream(lexer)
    const parser = new SourCParser(tokenStream)
    parser.buildParseTree = true
    try {
      const tree = parser.translationUnit()
      program = convertSource(tree)
    } catch (error) {
      if (error instanceof FatalSyntaxError) {
        context.errors.push(error)
      } else {
        throw error
      }
    }
    const hasErrors = context.errors.find(m => m.severity === ErrorSeverity.ERROR)
    if (program && !hasErrors) {
      console.log(JSON.stringify(program, null, 2))
      return program
    } else {
      return undefined
    }
  } else {
    return undefined
  }
}

/*
import { CalcLexer } from '../lang/CalcLexer'
import {
  AdditionContext,
  CalcParser,
  DivisionContext,
  ExpressionContext,
  MultiplicationContext,
  NumberContext,
  ParenthesesContext,
  PowerContext,
  StartContext,
  SubtractionContext
} from '../lang/CalcParser'
import { CalcVisitor } from '../lang/CalcVisitor'

class ExpressionGenerator implements CalcVisitor<es.Expression> {
  visitNumber(ctx: NumberContext): es.Expression {
    return {
      type: 'Literal',
      value: parseInt(ctx.text),
      raw: ctx.text,
      loc: contextToLocation(ctx)
    }
  }
  visitParentheses(ctx: ParenthesesContext): es.Expression {
    return this.visit(ctx.expression())
  }
  visitPower(ctx: PowerContext): es.Expression {
    return {
      type: 'BinaryExpression',
      operator: '^',
      left: this.visit(ctx._left),
      right: this.visit(ctx._right),
      loc: contextToLocation(ctx)
    }
  }

  visitMultiplication(ctx: MultiplicationContext): es.Expression {
    return {
      type: 'BinaryExpression',
      operator: '*',
      left: this.visit(ctx._left),
      right: this.visit(ctx._right),
      loc: contextToLocation(ctx)
    }
  }
  visitDivision(ctx: DivisionContext): es.Expression {
    return {
      type: 'BinaryExpression',
      operator: '/',
      left: this.visit(ctx._left),
      right: this.visit(ctx._right),
      loc: contextToLocation(ctx)
    }
  }
  visitAddition(ctx: AdditionContext): es.Expression {
    return {
      type: 'BinaryExpression',
      operator: '+',
      left: this.visit(ctx._left),
      right: this.visit(ctx._right),
      loc: contextToLocation(ctx)
    }
  }

  visitSubtraction(ctx: SubtractionContext): es.Expression {
    return {
      type: 'BinaryExpression',
      operator: '-',
      left: this.visit(ctx._left),
      right: this.visit(ctx._right),
      loc: contextToLocation(ctx)
    }
  }

  visitExpression?: ((ctx: ExpressionContext) => es.Expression) | undefined
  visitStart?: ((ctx: StartContext) => es.Expression) | undefined

  visit(tree: ParseTree): es.Expression {
    return tree.accept(this)
  }
  visitChildren(node: RuleNode): es.Expression {
    const expressions: es.Expression[] = []
    for (let i = 0; i < node.childCount; i++) {
      expressions.push(node.getChild(i).accept(this))
    }
    return {
      type: 'SequenceExpression',
      expressions
    }
  }
  visitTerminal(node: TerminalNode): es.Expression {
    return node.accept(this)
  }

  visitErrorNode(node: ErrorNode): es.Expression {
    throw new FatalSyntaxError(
      {
        start: {
          line: node.symbol.line,
          column: node.symbol.charPositionInLine
        },
        end: {
          line: node.symbol.line,
          column: node.symbol.charPositionInLine + 1
        }
      },
      `invalid syntax ${node.text}`
    )
  }
}

function convertExpression(expression: ExpressionContext): es.Expression {
  const generator = new ExpressionGenerator()
  return expression.accept(generator)
}
*/
