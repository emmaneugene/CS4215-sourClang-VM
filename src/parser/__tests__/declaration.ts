import createContext from '../../createContext'
import { Context, Variant } from '../../types'
import { parse } from '../parser'

const f = (when: string, should: string) => `When_\'${when}\'_Should_${should}`

let context: Context = createContext(Variant.DEFAULT, undefined, undefined)

beforeAll(() => {
  console.log = () => {}
})

beforeEach(() => {
  context = createContext(Variant.DEFAULT, undefined, undefined)
})

const CHAR_TEST_CASES: string[] = [
  'char x = 32767;',
  'unsigned char x = 32767;',
  'signed char x = 32767;',
  'char * x = 32767;',
  'unsigned char * x = 32767;',
  'signed char ** x = 32767;'
]

describe('Char', () => {
  for (const source of CHAR_TEST_CASES) {
    it(f(source, 'ParseSuccessfully'), () => {
      const result = parse(source, context)
      expect(result).toMatchSnapshot()
    })
  }
})

const SHORT_TEST_CASES: string[] = [
  'short x = 65535;',
  'unsigned short x = 65535;',
  'signed short x = 65535;',
  'short * x = 65535;',
  'unsigned short ** x = 65535;',
  'signed short * x = 65535;'
]

describe('Short', () => {
  for (const source of SHORT_TEST_CASES) {
    it(f(source, 'ParseSuccessfully'), () => {
      const result = parse(source, context)
      expect(result).toMatchSnapshot()
    })
  }
})

const INT_TEST_CASES: string[] = [
  'int x = 65535;',
  'unsigned int x = 65535;',
  'signed int x = 65535;',
  'int * * x = 65535;',
  'unsigned int * x = 65535;',
  'signed int * x = 65535;'
]

describe('Int', () => {
  for (const source of INT_TEST_CASES) {
    it(f(source, 'ParseSuccessfully'), () => {
      const result = parse(source, context)
      expect(result).toMatchSnapshot()
    })
  }
})

const LONG_TEST_CASES: string[] = [
  'long x = 65535;',
  'unsigned long x = 65535;',
  'signed long x = 65535;',
  'long * x = 65535;',
  'unsigned long * * x = 65535;',
  'signed long * x = 65535;'
]

describe('Long', () => {
  for (const source of LONG_TEST_CASES) {
    it(f(source, 'ParseSuccessfully'), () => {
      const result = parse(source, context)
      expect(result).toMatchSnapshot()
    })
  }
})

const FLOAT_TEST_CASES: string[] = ['float x = 65535;', 'float *** x = 65535;']

describe('Float', () => {
  for (const source of FLOAT_TEST_CASES) {
    it(f(source, 'ParseSuccessfully'), () => {
      const result = parse(source, context)
      expect(result).toMatchSnapshot()
    })
  }
})

const DOUBLE_TEST_CASES: string[] = ['double x = 65535;', 'double * x = 65535;']

describe('Double', () => {
  for (const source of DOUBLE_TEST_CASES) {
    it(f(source, 'ParseSuccessfully'), () => {
      const result = parse(source, context)
      expect(result).toMatchSnapshot()
    })
  }
})

const FAIL_TEST_CASE: string[] = [
  'unsigned float x = 1;',
  'unsigned double x = 1;',
  'unsigned unsigned * x = 1;',
  'unsigned signed x = 1;',
  'signed unsigned x = 1;',
  'signed signed x = 1;',
  'char char x = 1;',
  'short short x = 1;',
  'int int x = 1;',
  'long long x = 1;',
  'unsigned signed long x = 1;'
]

describe('Bad Syntax', () => {
  for (const source of FAIL_TEST_CASE) {
    it(f(source, 'ThrowError'), () => {
      expect(() => parse(source, context)).toThrow()
    })
  }
})
