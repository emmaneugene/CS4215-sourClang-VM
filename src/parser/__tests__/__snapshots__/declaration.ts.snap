// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Char When_'char * x = 32767;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "char",
                "name": "x",
                "pointerList": Array [
                  "*",
                  undefined,
                ],
                "type": "Identifier",
              },
              "init": Object {
                "raw": "32767",
                "type": "Literal",
                "value": 32767,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Char When_'char x = 32767;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "char",
                "name": "x",
                "pointerList": undefined,
                "type": "Identifier",
              },
              "init": Object {
                "raw": "32767",
                "type": "Literal",
                "value": 32767,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Char When_'signed char ** x = 32767;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "char",
                "name": "x",
                "pointerList": Array [
                  "*",
                  Array [
                    "*",
                    undefined,
                  ],
                ],
                "type": "Identifier",
              },
              "init": Object {
                "raw": "32767",
                "type": "Literal",
                "value": 32767,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Char When_'signed char x = 32767;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "char",
                "name": "x",
                "pointerList": undefined,
                "type": "Identifier",
              },
              "init": Object {
                "raw": "32767",
                "type": "Literal",
                "value": 32767,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Char When_'unsigned char * x = 32767;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "unsigned_char",
                "name": "x",
                "pointerList": Array [
                  "*",
                  undefined,
                ],
                "type": "Identifier",
              },
              "init": Object {
                "raw": "32767",
                "type": "Literal",
                "value": 32767,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Char When_'unsigned char x = 32767;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "unsigned_char",
                "name": "x",
                "pointerList": undefined,
                "type": "Identifier",
              },
              "init": Object {
                "raw": "32767",
                "type": "Literal",
                "value": 32767,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Double When_'double * x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "double",
                "name": "x",
                "pointerList": Array [
                  "*",
                  undefined,
                ],
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Double When_'double x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "double",
                "name": "x",
                "pointerList": undefined,
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Float When_'float *** x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "float",
                "name": "x",
                "pointerList": Array [
                  "*",
                  Array [
                    "*",
                    Array [
                      "*",
                      undefined,
                    ],
                  ],
                ],
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Float When_'float x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "float",
                "name": "x",
                "pointerList": undefined,
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Int When_'int * * x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "int",
                "name": "x",
                "pointerList": Array [
                  "*",
                  Array [
                    "*",
                    undefined,
                  ],
                ],
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Int When_'int x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "int",
                "name": "x",
                "pointerList": undefined,
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Int When_'signed int * x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "int",
                "name": "x",
                "pointerList": Array [
                  "*",
                  undefined,
                ],
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Int When_'signed int x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "int",
                "name": "x",
                "pointerList": undefined,
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Int When_'unsigned int * x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "unsigned_int",
                "name": "x",
                "pointerList": Array [
                  "*",
                  undefined,
                ],
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Int When_'unsigned int x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "unsigned_int",
                "name": "x",
                "pointerList": undefined,
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Long When_'long * x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "long",
                "name": "x",
                "pointerList": Array [
                  "*",
                  undefined,
                ],
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Long When_'long x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "long",
                "name": "x",
                "pointerList": undefined,
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Long When_'signed long * x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "long",
                "name": "x",
                "pointerList": Array [
                  "*",
                  undefined,
                ],
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Long When_'signed long x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "long",
                "name": "x",
                "pointerList": undefined,
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Long When_'unsigned long * * x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "unsigned_long",
                "name": "x",
                "pointerList": Array [
                  "*",
                  Array [
                    "*",
                    undefined,
                  ],
                ],
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Long When_'unsigned long x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "unsigned_long",
                "name": "x",
                "pointerList": undefined,
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Short When_'short * x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "short",
                "name": "x",
                "pointerList": Array [
                  "*",
                  undefined,
                ],
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Short When_'short x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "short",
                "name": "x",
                "pointerList": undefined,
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Short When_'signed short * x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "short",
                "name": "x",
                "pointerList": Array [
                  "*",
                  undefined,
                ],
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Short When_'signed short x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "short",
                "name": "x",
                "pointerList": undefined,
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Short When_'unsigned short ** x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "unsigned_short",
                "name": "x",
                "pointerList": Array [
                  "*",
                  Array [
                    "*",
                    undefined,
                  ],
                ],
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Short When_'unsigned short x = 65535;'_Should_ParseSuccessfully 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "declarations": Array [
            Object {
              "id": Object {
                "datatype": "unsigned_short",
                "name": "x",
                "pointerList": undefined,
                "type": "Identifier",
              },
              "init": Object {
                "raw": "65535",
                "type": "Literal",
                "value": 65535,
              },
              "type": "VariableDeclarator",
            },
          ],
          "kind": "var",
          "type": "VariableDeclaration",
        },
      ],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;
